src/front/wgsl/number.rs

```diff
@@ -0,0 +1,506 @@
+use std::borrow::Cow;
+
+use super::{NumberError, Token};
+
+/// When using this type assume no Abstract Int/Float for now
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum Number {
+    /// Abstract Int (-2^63 ≤ i < 2^63)
+    Int(i64),
+    /// Abstract Float (IEEE-754 binary64)
+    Float(f64),
+    /// Concrete i32
+    I32(i32),
+    /// Concrete u32
+    U32(u32),
+    /// Concrete f32
+    F32(f32),
+}
+
+// TODO: when implementing Creation-Time Expressions, remove the ability to match the minus sign
+
+pub(super) fn consume_number(input: &str) -> (Token<'_>, &str) {
+    let res = if let Some(ExtractInfo { info, rest }) = extract_info(input) {
+        (map(info), rest)
+    } else {
+        (Err(NumberError::Invalid), input)
+    };
+
+    let num = match res.0 {
+        Ok(Number::Int(num)) => {
+            use std::convert::TryFrom;
+            i32::try_from(num)
+                .map(Number::I32)
+                .map_err(|_| NumberError::NotRepresentable)
+        }
+        Ok(Number::Float(num)) => {
+            let num = num as f32;
+            if num.is_finite() {
+                Ok(Number::F32(num))
+            } else {
+                Err(NumberError::NotRepresentable)
+            }
+        }
+        num => num,
+    };
+    (Token::Number(num), res.1)
+}
+
+struct ExtractInfo<'a> {
+    info: Info<'a>,
+    rest: &'a str,
+}
+
+enum Info<'a> {
```

Comment:
I don't think this enum is earning its keep. It looks to me as if you could:
1) replace each variant with a function (i.e. `parse_hex_float_missing_exponent`) whose arguments are the variant's fields
2) call those functions from `extract_info` instead of constructing the variants, and
3) eliminate the `map` function altogether.

The comments on the variants here would make sense as doc comments on the functions.